xds/internal/xdsclient/callback_serializer_test.go

```diff
@@ -0,0 +1,228 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package xdsclient
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/grpc/internal/grpcsync"
+)
+
+// TestCallbackSerializer_Schedule_FIFO verifies that callbacks are executed in
+// the same order in which they were scheduled.
+func (s) TestCallbackSerializer_Schedule_FIFO(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	cs := newCallbackSerializer(ctx)
+	defer cancel()
+
+	// We have two channels, one to record the order of scheduling, and the
+	// other to record the order of execution. We spawn a bunch of goroutines
+	// which record the order of scheduling and call the actual Schedule()
+	// method as well.  The callbacks record the order of execution.
+	//
+	// We need to grab a lock to record order of scheduling to guarantee that
+	// the act of recording and the act of calling Schedule() happen atomically.
+	const numCallbacks = 100
+	var mu sync.Mutex
+	scheduleOrderCh := make(chan int, numCallbacks)
+	executionOrderCh := make(chan int, numCallbacks)
+	for i := 0; i < numCallbacks; i++ {
+		go func(id int) {
+			mu.Lock()
+			defer mu.Unlock()
+			scheduleOrderCh <- id
+			cs.Schedule(func(ctx context.Context) {
+				select {
+				case <-ctx.Done():
+					return
+				case executionOrderCh <- id:
+				}
+			})
+		}(i)
+	}
+
+	// Spawn a couple of goroutines to capture the order or scheduling and the
+	// order of execution.
+	scheduleOrder := make([]int, numCallbacks)
+	executionOrder := make([]int, numCallbacks)
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		for i := 0; i < numCallbacks; i++ {
+			select {
+			case <-ctx.Done():
+				return
+			case id := <-scheduleOrderCh:
+				scheduleOrder[i] = id
+			}
+		}
+	}()
+	go func() {
+		defer wg.Done()
+		for i := 0; i < numCallbacks; i++ {
+			select {
+			case <-ctx.Done():
+				return
+			case id := <-executionOrderCh:
+				executionOrder[i] = id
+			}
+		}
+	}()
+	wg.Wait()
+
+	if diff := cmp.Diff(executionOrder, scheduleOrder); diff != "" {
+		t.Fatalf("Callbacks are not executed in scheduled order. diff(-want, +got):\n%s", diff)
+	}
+}
+
+// TestCallbackSerializer_Schedule_RunToCompletion verifies that all scheduled
+// callbacks are run to completion.
```

Comment:
The objective of this test is to ensure that all scheduled callbacks are executed. The objective of the previous test is to ensure that the callbacks are executed in the order in which they were scheduled. Technically, what is being tested here is also tested in the previous test. But the previous test is more complicated than this one because it needs to determine the scheduling order. So, if there is something broken only in the logic which ensures that all scheduled callbacks are executed, this test will be easier to use for debugging.

Reply:
It seems like the only difference is the previous test does not sort the expected/actual callbacks executed.  If you want to keep it to make debugging more obvious, maybe refactor into a helper so the rest of the test isn't duplicated?

Reply:
The first test calls `Schedule()` while holding a lock. So, in the first test, calls to `Schedule()` happen one after the other. We need to guarantee this because that is the only way we can ensure that callbacks are executed in the order in which they are scheduled.

The second test calls `Schedule()` without holding a lock. So, this is where calls are `Schedule()` are happening in a truly concurrent fashion. And I feel it is important that we retain that test.

As far as refactoring them to weed out repeated code is concerned, the only thing that can be factored out into a common function is the logic where a couple of goroutines are spawned to capture the order of scheduling and execution. I tried making a function for this, but it was very ugly because of the number of arguments that need to be passed and the result parameters. 

Let me know if you think otherwise. Thanks.

Reply:
Maybe simplify this test a bit further?  Why capture the execution order if you don't verify it?  How about just a waitgroup and make each scheduled callback execute `wg.Done()`?  If they are all executed, then `wg.Wait()` will eventually unblock and the test can pass.

Also maybe rename this to `TestCallbackSerializer_Schedule_ConcurrentCalls()` or something?

Reply:
Done. Thanks.